src/services/data-targets/fiware-data-target.service.ts

```diff
@@ -9,9 +10,53 @@ import { DataTargetSendStatus } from "@interfaces/data-target-send-status.interf
 import { FiwareDataTargetConfiguration } from "@interfaces/fiware-data-target-configuration.interface";
 import { BaseDataTargetService } from "@services/data-targets/base-data-target.service";
 
+type TokenEndpointResponse = {
+    access_token: string,
+    expires_in: number,
+}
+
+@Injectable()
+export class AuthenticationTokenProvider {
+
+    private readonly logger = new Logger(AuthenticationTokenProvider.name);
+
+    constructor(private httpService: HttpService, @Inject(CACHE_MANAGER) private cacheManager: Cache) {
+    }
+
+    async getToken(config: FiwareDataTargetConfiguration): Promise<string> {
+        const key = config.clientId + config.clientSecret;
+        const token = await this.cacheManager.get<string>(key)
+        if (token) {
+            this.logger.debug('Token found')
+            return token
+        } else {
+            try {
+                const buffer = Buffer.from(`${config.clientId}:${config.clientSecret}`);
+                const encodedCredentials = buffer.toString('base64');
+                const params = new URLSearchParams([['grant_type', 'client_credentials']])
+                const { data }: { data: TokenEndpointResponse } = await this.httpService.post(config.tokenEndpoint, params, {
+                    headers: {
+                        'Authorization': `Basic ${encodedCredentials}`,
+                        'Content-Type': 'application/x-www-form-urlencoded',
+                    },
+                }).toPromise()
+
+                const clockSkew = 30
+                const ttl = data.expires_in - clockSkew
+                this.logger.debug(`AuthenticationTokenProvider caching token for ${config.clientId} (expires in ${ttl} seconds)`)
+                this.cacheManager.set(key, data.access_token, { ttl })
+                return data.access_token
+            }
+            catch (err) {
+                this.logger.error(`AuthenticationTokenProvider got error ${err}`)
```

Comment:
It returns `undefined` on error. This is okay if you update the return type. If you do that, remember to handle undefined when fetching the token.